### PR TITLE
Libsharp installation issues

### PR DESCRIPTION
**Pull request by damonge**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

closes #26
closes #23

### CODE REVIEW

**Comment by beckermr** [comment]
What is this or doing?

**Comment by beckermr** [comment]
You cannot vendor libsharp. You have to install it from conda.

**Comment by damonge** [comment]
Right, sorry @beckermr .
I'm trying to build namaster through `setup.py`, which downloads and compiles libsharp2. Why is this not possible? How would this be different to compiling my own internal C library?

**Comment by beckermr** [comment]
I see. If you go this route, you need to include the license files for libsharp2 and change the recipe metadata. You also need to ensure that you are not installing extra .so's etc. Instead, you need to directly include the source in your C extension and compile libsharp2 with your project directly.

**Comment by damonge** [comment]
OK, thanks a lot.
I think it just may be easier to revert to libsharp... I'll play around with this a bit before I give up. Don't mind me, and I'll ping you if I actually need a review.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -26,20 +24,20 @@ requirements:
...
   host:
     - python
     - pip
     - gsl
     - fftw
     - cfitsio
-    - libsharp
     - chealpix
     - numpy
+    - libtool
```
Comment (beckermr) [inline recipe/meta.yaml]: Just an FYI, a build tool like this never goes in host.

Comment (damonge) [reply recipe/meta.yaml]: Ah, thanks a lot. I wasn't sure from the description I found online.

**Comment by damonge** [comment]
@beckermr sorry for the intermittent noise. This should be ready for review.

**Comment by beckermr** [comment]
A few questions / comments:

1. How are you vendoring libsharp2?
2. You need to include the libsharp2 license in the package.

**Comment by damonge** [comment]
Sorry for the delay (and thank you!):
1. Not sure what "vendoring" means in this context. My aim is to just compile the libsharp static library and then namaster with it. Does that help?
2. Where should I put the license? And why is this not necessary for all the other dependencies if we're not shipping those binaries?

**Comment by beckermr** [comment]
1. By vendoring I mean including the libsharp code in with this package. 
2. The conda deps listed in run themselves are not shipped with your code. However, in this case you are adding in source/compiled code from libsharp into your binary and that is getting put into the namaster package. Once you do that, you need the license. Any other deps treated in the same way need their licenses too. 

You can put the license in the recipe dir here and then change the recipe license key to a list of licenses.

**Comment by damonge** [comment]
OK, thanks for the clarification.
As far as I understand I'm not including the libsharp2 code, but only using it to compile namaster. Does compiled code pose a problem?
I'll add the license ASAP, thanks.

**Comment by beckermr** [comment]
> As far as I understand I'm not including the libsharp2 code, but only using it to compile namaster. Does compiled code pose a problem?

Compiled code still needs a license yes.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -26,38 +24,41 @@ requirements:
...
+    - pytest -vv pymaster/tests
 
 about:
   home: https://github.com/LSSTDESC/NaMaster
   license: BSD-3-Clause
   license_family: BSD
-  license_file: LICENSE
+  license_file:
+    - LICENSE
+    - COPYING
```
Comment (beckermr) [inline recipe/meta.yaml:65]: You need to commit this file from libsharp2 to the feedstock at namaster-feedstock/recipe/COPYING.

You also need to adjust the license indicator to reflect the license of libsharp2 as well.

**Review by beckermr — CHANGES_REQUESTED** [review]

File: recipe/meta.yaml (hunk 2 above)
Comment (damonge) [reply recipe/meta.yaml:65]: Oops, forgot to include that. Done!

Comment (damonge) [reply recipe/meta.yaml:65]: > You also need to adjust the license indicator to reflect the license of libsharp2 as well.

Actually, I just realized I'm not sure how to do this, and I couldn't find anything online.

But this leads me to a more basic question (which maybe I should have started with). libsharp is GPL-2, whereas namaster is BSD-3. I think they are compatible (I believe the problem is with GPL-3), so is this a problem?

File: recipe/meta.yaml [hunk 3]
```diff
@@ -54,7 +54,9 @@ test:
 
 about:
   home: https://github.com/LSSTDESC/NaMaster
-  license: BSD-3-Clause
+  license:
+    - BSD-3-Clause
+    - GPL-2
```
Comment (beckermr) [inline recipe/meta.yaml]:
```suggestion
  license: BSD-3-Clause AND GPL-2.0-only
```

File: recipe/meta.yaml (hunk 2 above)
Comment (beckermr) [reply recipe/meta.yaml:65]: Seems fine.

File: recipe/meta.yaml (hunk 3 above)
Comment (damonge) [reply recipe/meta.yaml]: Thanks so much!

File: recipe/meta.yaml (hunk 2 above)
Comment (damonge) [reply recipe/meta.yaml:65]: Thanks!

**Review by beckermr — CHANGES_REQUESTED** [review]
We need to make sure we use the conda-forge linking and compilation flags everywhere: https://github.com/LSSTDESC/NaMaster/blob/master/scripts/install_libsharp.sh#L26

**Comment by damonge** [comment]
OK, thanks. What are those? I couldn't find anything special here: https://github.com/conda-forge/namaster-feedstock/blob/master/recipe/build.sh

**Comment by damonge** [comment]
Or do you mean we should make sure conda-forge uses the CFLAGS that are currently under Travis in the script you linked?

**Comment by beckermr** [comment]
They are in env vars CFLags, ldflags, etc.

**Comment by beckermr** [comment]
You cannot use march native here.

File: recipe/libsharp_compflags.patch [hunk 4]
```diff
@@ -0,0 +1,13 @@
+diff --git a/scripts/install_libsharp.sh b/scripts/install_libsharp.sh
+index fcfa0b7..eeabb84 100755
+--- a/scripts/install_libsharp.sh
++++ b/scripts/install_libsharp.sh
+@@ -23,7 +23,7 @@ if [[ $TRAVIS ]] ; then
+     CFLAGS="-std=c99 -O3 -ffast-math"
+ else
+     echo "Using -march=native. Binary will not be portable."
+-    CFLAGS="-march=native -std=c99 -O3 -ffast-math"
++    CFLAGS+="-march=native -std=c99 -O3 -ffast-math"
```
Comment (beckermr) [inline recipe/libsharp_compflags.patch]:
```suggestion
+    CFLAGS+="-std=c99 -O3 -ffast-math"
```

Comment (damonge) [reply recipe/libsharp_compflags.patch]: of course, that makes sense!

**Comment by damonge** [comment]
Just checking, are we missing something else here?

**Comment by beckermr** [comment]
I want to double check the builds one more time. Just to confirm, you are statically linking libsharp2?

**Comment by damonge** [comment]
Yep, confirmed. This only generates and links the static libraries.

File: recipe/meta.yaml [hunk 5]
```diff
@@ -26,38 +25,41 @@ requirements:
...
+    - libtool
   host:
     - python
     - pip
     - gsl
     - fftw
     - cfitsio
-    - libsharp
     - chealpix
     - numpy
```
Comment (beckermr) [inline recipe/meta.yaml:36]:
```suggestion
    - numpy
    - libblas
    - libcblas
```

**Comment by beckermr** [comment]
@conda-forge-admin rerender

File: recipe/meta.yaml [hunk 6]
```diff
@@ -26,38 +25,43 @@ requirements:
...
+    - libblas
+    - libcblas
```
Comment (beckermr) [inline recipe/meta.yaml:38]:
```suggestion
    - libcblas
    - liblapack
```

**Comment by beckermr** [comment]
@conda-forge-admin rerender

**Comment by beckermr** [comment]
Which linear algebra libraries do you use? I think this has been a bug in the recipe for a long time now.

**Comment by damonge** [comment]
All the linear algebra is done via gsl (gsl and gslcblas). Why do you think this has been a bug?

**Comment by beckermr** [comment]
I think we need the underlying SOs explicitly in host to get things setup right. 

see here: https://conda-forge.org/docs/maintainer/knowledge_base.html?highlight=blas#blas

**Review by beckermr — APPROVED** [review]

File: recipe/meta.yaml [hunk 7]
```diff
@@ -8,10 +8,9 @@ package:
...
   url: https://github.com/LSSTDESC/NaMaster/archive/v{{ version }}.tar.gz
-  sha256: 60ceee5fd971c7fc081ca278ef91ea07ed4debf991e6360bc2f0a3a5ca2663c8
+  sha256: 05a676010a7e0b860af845bea8e98e05f0d94b36498c8b63290bdc8c5ef97499
   patches:
-    - 0000_nmt_test_field.patch
-    - healpy_test_fixes_0000.patch
+    - libsharp_compflags.patch
 
 build:
   number: 2
```
Comment (beckermr) [inline recipe/meta.yaml]:
```suggestion
  number: 0
```

**Comment by beckermr** [comment]
@conda-forge-admin rerender

**Pull request merged**